File: leetcode/0819-most-common-word/mostCommonWord.go
package leetcode

import (
	"regexp"
	"strings"
)

func mostCommonWord(paragraph string, banned []string) string {
	var (
		result   = ""
		maxCount = 0
	)

	var bannedMap = map[string]struct{}{}
	var wordCountMap = map[string]int{}
	var re = regexp.MustCompile(`[!?',;.]`)
	var words = strings.Fields(re.ReplaceAllString(strings.ToLower(paragraph), " "))

	for _, string := range banned {
		bannedMap[string] = struct{}{}
	}

	for _, word := range words {
		if _, banned := bannedMap[word]; banned {
			continue
		}

		if count, wordExists := wordCountMap[word]; wordExists {
			wordCountMap[word] = count + 1
			continue
		}

		wordCountMap[word] = 1
	}

	for word, count := range wordCountMap {
		if count <= maxCount {
			continue
		}

		maxCount = count
		result = word
	}

	return result
}

File: leetcode/0819-most-common-word/mostCommonWord.ts
function mostCommonWord(paragraph: string, banned: string[]): string {
  let maxWordCount: [string, number] = ["", 0];

  const words = paragraph.toLowerCase().replace(/[!?',;.]/g, ' ').split(" ").filter((item) => item != "");
  const wordCountMap = new Map<string, number>();

  for (const word of words) {
    if (banned.includes(word)) {
      continue;
    }
    if (wordCountMap.has(word)) {
      wordCountMap.set(word, wordCountMap.get(word)! + 1);
      continue;
    }

    wordCountMap.set(word, 1);
  }

  for (const [word, count] of wordCountMap) {
    if (count <= maxWordCount[1]!) {
      continue;
    }

    maxWordCount = [word, count];
  }

  return maxWordCount[0]!;
}

export default mostCommonWord;
